File: apps/api/signalyze_api/engine/signals/burst_ratio.py
"""Twin of ``signals/burst-ratio.ts``."""

from __future__ import annotations

from collections.abc import Sequence

from signalyze_api.engine.data_files import THRESHOLDS
from signalyze_api.engine.mathutil import clamp01, iso_from_day, ramp, round6
from signalyze_api.engine.prepare import PreparedReview
from signalyze_api.models import SignalResult

T = THRESHOLDS.burst_ratio


def burst_ratio(reviews: Sequence[PreparedReview]) -> SignalResult:
    """Share of reviews inside the busiest 14-day window, compared with an even flow.

    value = peak share.
    """
    days = sorted(r.day for r in reviews)
    n = len(days)
    first = days[0]
    last = days[n - 1]
    lifespan_days = last - first + 1

    peak = 0
    peak_start = first
    lo = 0
    for hi in range(n):
        while days[hi] - days[lo] >= T.windowDays:
            lo += 1
        count = hi - lo + 1
        if count > peak:
            peak = count
            peak_start = days[lo]
    peak_share = peak / n
    expected_share = 1.0 if lifespan_days <= T.windowDays else T.windowDays / lifespan_days
    excess = (
        0.0
        if expected_share >= 1
        else clamp01((peak_share - expected_share) / (1 - expected_share))
    )
    unusualness = ramp(excess, T.excessLow, T.excessHigh)

    return SignalResult(
        id="burst_ratio",
        unusualness=round6(unusualness),
        value=round6(peak_share),
        details={
            "peakShare": round6(peak_share),
            "peakCount": peak,
            "peakWindowStart": iso_from_day(peak_start),
            "peakWindowEnd": iso_from_day(peak_start + T.windowDays - 1),
            "lifespanDays": lifespan_days,
            "expectedShare": round6(expected_share),
            "excess": round6(excess),
            "windowDays": T.windowDays,
        },
        available=True,
    )

File: packages/signals/src/signals/burst-ratio.ts
import type { SignalResult } from '@signalyze/shared';
import thresholds from '../../data/thresholds.json';
import { clamp01, ramp, round6 } from '../math';
import type { PreparedReview } from '../prepare';

const T = thresholds.burst_ratio;

/**
 * Share of reviews inside the busiest 14-day window, compared with the share a
 * perfectly even flow would put there (window / lifespan). value = peak share.
 */
export function burstRatio(reviews: readonly PreparedReview[]): SignalResult {
  const days = reviews.map((r) => r.day).sort((a, b) => a - b);
  const n = days.length;
  const first = days[0]!;
  const last = days[n - 1]!;
  const lifespanDays = last - first + 1;

  let peak = 0;
  let peakStart = first;
  let lo = 0;
  for (let hi = 0; hi < n; hi += 1) {
    while (days[hi]! - days[lo]! >= T.windowDays) lo += 1;
    const count = hi - lo + 1;
    if (count > peak) {
      peak = count;
      peakStart = days[lo]!;
    }
  }
  const peakShare = peak / n;
  const expectedShare = lifespanDays <= T.windowDays ? 1 : T.windowDays / lifespanDays;
  const excess =
    expectedShare >= 1 ? 0 : clamp01((peakShare - expectedShare) / (1 - expectedShare));
  const unusualness = ramp(excess, T.excessLow, T.excessHigh);

  return {
    id: 'burst_ratio',
    unusualness: round6(unusualness),
    value: round6(peakShare),
    details: {
      peakShare: round6(peakShare),
      peakCount: peak,
      peakWindowStart: isoFromDay(peakStart),
      peakWindowEnd: isoFromDay(peakStart + T.windowDays - 1),
      lifespanDays,
      expectedShare: round6(expectedShare),
      excess: round6(excess),
      windowDays: T.windowDays,
    },
    available: true,
  };
}

function isoFromDay(day: number): string {
  return new Date(day * 86_400_000).toISOString().slice(0, 10);
}
